hat/app/org/hatdex/hat/api/controllers/RichData.scala

```diff
@@ -679,12 +679,71 @@ class RichData @Inject() (
           case Left(PublicKeyRequestFailure.InvalidPublicKeyFailure(failureDescription)) => {
             Left(ContractVerificationFailure.ServiceRespondedWithFailure(s"The Adjudicator Service responded with an error: ${failureDescription}"))
           }
-          case Right(PublicKeyReceived(publicKey)) => verifyJwyClaim(contractRequestBodyRefined, publicKey)
+          case Right(PublicKeyReceived(publicKey)) => verifyJwtClaim(contractRequestBodyRefined, publicKey)
         }
       }
     }
   }
 
+
+
+  /* *** Contract Data *** */
+
+  def getContractData(
+    namespace: String,
+    endpoint: String,
+    orderBy: Option[String],
+    ordering: Option[String],
+    skip: Option[Int],
+    take: Option[Int]): Action[ContractRequestBody] =
+    UserAwareAction.async(parsers.json[ContractRequestBody]) { implicit request =>
+      val contractRequestBody = request.body
+      val requestIsAllowed = assessRequest(contractRequestBody, namespace)
+
+      requestIsAllowed.flatMap { testResult =>
+        testResult match {
+          case Right(RequestVerified(ns)) => makeData(namespace, endpoint, orderBy, ordering, skip, take)
+          case _ => Future.successful(NotFound)
+        }
+      } recover {
+        case _ => NotFound
+      }
+    }
+
+
+  def assessRequest(
+    contractRequestBody: ContractRequestBody,
+    namespace: String): Future[Either[RequestValidationFailure, RequestVerified]] = {
+      val eventuallyMaybeDecision = verifyContract(contractRequestBody)
+      val eventuallyMaybeApp = trustedApplicationProvider.application(contractRequestBody.contractId)
+
+      eventuallyMaybeDecision.flatMap { maybeDecision =>
+        eventuallyMaybeApp.flatMap { maybeApp =>
+          decide(maybeDecision, maybeApp, namespace) match {
+            case Some(_ns) => Future.successful(Right(RequestVerified(s"Token: ${contractRequestBody.contractId}")))
+            case None => Future.successful(Left(RequestValidationFailure.InvalidShortLivedToken(s"Token: ${contractRequestBody.contractId}")))
+          }
+        }
+      } recover {
+        case _e => Left(RequestValidationFailure.InvalidShortLivedToken(s"Token: ${contractRequestBody.contractId}"))
+      }
+    }
+
+
+  def decide(eitherDecision: Either[ContractVerificationFailure, ContractVerificationSuccess], maybeApp: Option[Application], namespace: String): Option[String] = {
+    import ContractVerificationSuccess._
+
+    (eitherDecision, maybeApp) match {
+      case (Right(JwtClaimVerified(_jwtClaim)) , Some(app)) => {
+        verifyNamespace(app, namespace)
```

Comment:
Using an option here and then converting to a either seems a little backwards
The logic escapes the decide function (a public member) and the user of the method needs to figure out why a none was returned
Would it be better to return an either from this (and verifyNamespace) 
Then no conversion from option would be required ? It would become Future(decide(...))